onionshare/web.py

```diff
@@ -343,96 +344,104 @@ def upload_logic(slug_candidate=''):
             """
             Upload files.
             """
-            # Make sure downloads_dir exists
-            valid = True
-            try:
-                self.common.validate_downloads_dir()
-            except DownloadsDirErrorCannotCreate:
-                self.add_request(Web.REQUEST_ERROR_DOWNLOADS_DIR_CANNOT_CREATE, request.path)
-                print(strings._('error_cannot_create_downloads_dir').format(self.common.settings.get('downloads_dir')))
-                valid = False
-            except DownloadsDirErrorNotWritable:
-                self.add_request(Web.REQUEST_ERROR_DOWNLOADS_DIR_NOT_WRITABLE, request.path)
-                print(strings._('error_downloads_dir_not_writable').format(self.common.settings.get('downloads_dir')))
-                valid = False
-            if not valid:
-                flash('Error uploading, please inform the OnionShare user', 'error')
-                if self.common.settings.get('public_mode'):
-                    return redirect('/')
-                else:
-                    return redirect('/{}'.format(slug_candidate))
+            while self.can_upload:
```

Comment:
I don't quite understand what this while loop is doing. I searched the code for `self.can_upload` and it seems to that it never gets set to False, and it seems like all code paths return from the function before ending the first loop anyway. Am I missing something?

Reply:
No, I'm probably just stupid.

self.can_upload gets set to False in the GUI in  timeout_finished_should_stop_server(), when the timer hsa expired, as a way of saying 'the timer's run out, but there's an upload running, no-one else Can Upload (false)'